src/internal/pfsdb/repos.go

```diff
@@ -0,0 +1,363 @@
+package pfsdb
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/jackc/pgconn"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+)
+
+const (
+	getRepoAndBranches = "SELECT repo.id, repo.name, repo.type, repo.project_id, " +
+		"repo.description, array_agg(branch.proto) AS branches, repo.created_at, repo.updated_at, project.name AS proj_name FROM pfs.repos repo " +
+		"JOIN core.projects project ON repo.project_id = project.id " +
+		"LEFT JOIN collections.branches branch ON project.name || '/' || repo.name || '.' || repo.type = branch.idx_repo "
+	noBranches = "{NULL}"
+)
+
+// ErrRepoNotFound is returned by GetRepo() when a repo is not found in postgres.
+type ErrRepoNotFound struct {
+	Project string
+	Name    string
+	Type    string
+	ID      pachsql.ID
+}
+
+// Error satisfies the error interface.
+func (err ErrRepoNotFound) Error() string {
+	if id := err.ID; id != 0 {
+		return fmt.Sprintf("repo id=%d not found", id)
+	}
+	return fmt.Sprintf("repo %s/%s.%s not found", err.Project, err.Name, err.Type)
+}
+
+func (err ErrRepoNotFound) GRPCStatus() *status.Status {
+	return status.New(codes.NotFound, err.Error())
+}
+
+func IsErrRepoNotFound(err error) bool {
+	return errors.As(err, &ErrRepoNotFound{})
+}
+
+// ErrRepoAlreadyExists is returned by CreateRepo() when a repo with the same name already exists in postgres.
+type ErrRepoAlreadyExists struct {
+	Project string
+	Name    string
+	Type    string
+}
+
+// Error satisfies the error interface.
+func (err ErrRepoAlreadyExists) Error() string {
+	if n, t := err.Name, err.Type; n != "" && t != "" {
+		return fmt.Sprintf("repo %s.%s already exists", n, t)
+	}
+	return "repo already exists"
+}
+
+func (err ErrRepoAlreadyExists) GRPCStatus() *status.Status {
+	return status.New(codes.AlreadyExists, err.Error())
+}
+
+func IsErrRepoAlreadyExists(err error) bool {
+	targetErr := &pgconn.PgError{}
+	ok := errors.As(err, targetErr)
```

Comment:
This is *very* surprising.  There is a type called `ErrRepoAlreadyExists` and this would return false for it.

Maybe an `IsDuplicateKeyErr` in `pachsql` would help deal with this other usecase.